src/silo/query_engine/actions/fasta.cpp

```diff
@@ -162,87 +170,147 @@ void Fasta::addSequencesToResultsForPartition(
    const {
    for (const std::string& sequence_name : sequence_names) {
       CHECK_SILO_QUERY(
          database.unaligned_nuc_sequences.contains(sequence_name),
          "Database does not contain an unaligned sequence with name: '" + sequence_name + "'"
-      )
+      );
    }
 
-   const std::string& primary_key_column = database.database_config.schema.primary_key;
-
-   size_t total_count = 0;
-   for (auto& filter : bitmap_filter) {
-      total_count += filter->cardinality();
-   }
-   CHECK_SILO_QUERY(
-      total_count <= SEQUENCE_LIMIT,
-      fmt::format("Fasta action currently limited to {} sequences", SEQUENCE_LIMIT)
+   size_t current_partition = 0;
+   // The unprocessed part of the result row numbers, counting from 0
+   // from the start of the current partition:
+   std::optional<Range<uint32_t>> remaining_result_row_indices;
+   return QueryResult::fromGenerator(
+      [sequence_names = sequence_names,
+       remaining_result_row_indices,
+       bitmap_filter = std::make_shared<std::vector<OperatorResult>>(std::move(bitmap_filter)),
+       &database,
+       current_partition](std::vector<QueryResultEntry>& results) mutable {
+         while (current_partition < database.partitions.size()) {
+            // We drain bitmap_filter as we process the query (because
+            // the only way to get an iterator over a bitmap is from the
+            // beginning?)! And `remaining_result_row_indices` is really
+            // only for show now.
+            auto& bitmap = (*bitmap_filter)[current_partition];
+            if (!remaining_result_row_indices.has_value()) {
+               // We set `remaining_result_row_indices` only once using the
+               // original, undrained bitmap.
+               remaining_result_row_indices = {
+                  {0, boost::numeric_cast<uint32_t, uint64_t>(bitmap->cardinality())}
+               };
+            }
+
+            const Range<uint32_t> result_row_indices =
+               remaining_result_row_indices->take(PARTITION_CHUNK_SIZE);
+            if (!result_row_indices.isEmpty()) {
+               SPDLOG_TRACE(
+                  "Fasta::execute: refill QueryResult for partition_index {}/{}, {}",
+                  current_partition,
+                  database.partitions.size(),
+                  result_row_indices.toString()
+               );
+               const std::string& primary_key_column = database.database_config.schema.primary_key;
+               const auto& database_partition = database.partitions[current_partition];
+
+               const uint32_t last_row_id = addSequencesToResultsForPartition(
+                  sequence_names,
+                  results,
+                  database_partition,
+                  bitmap,
+                  primary_key_column,
+                  result_row_indices.size()
+               );
+
+               assert(results.size() == result_row_indices.size());
```

Comment:
Let's `PANIC` instead?

Reply:
asserts have the advantage of no runtime overhead

Reply:
I've added issue #505 to do this, as there is pre-existing assert use in the code.

`assert` *does* have runtime overhead; some project instead uses an `assert` that is a noop on release builds, you're probably thinking of that (it also doesn't check anything then of course). Also some static code analysis tools can analyze some kinds of asserts at compile time (but I think they use their own syntax usually?); clang/g++ won't be able to do this in the general case.

Reply:
In release builds, source files are compiled with `-DNDEBUG`. [This transforms any `assert` into no-ops](https://en.cppreference.com/w/cpp/error/assert) and which are optimized away

Reply:
Ah, I didn't realize cassert did that itself! Thanks for the correction. I guess you'd like me to add an `#ifdef NDEBUG` then?

Reply:
(Let's move further discussion to #505.)